### PR TITLE
Treat TensorFlow output as non-batched.

### PR DESCRIPTION
**Pull request by michaelgsharp**
Currently, if TensorFlow has a -1 as the first dimension in an output tensor then ML.NET always treats it as the batch number. There are some cases where we don't want it treated as a batch number, but an actual dimension. This PR gives the user the ability to specify whether its batch output or not but keeps the default behavior of assuming it is batched.

This issue originally stems from #5364, though there was a lot of offline conversation about it.

### CODE REVIEW

File: src/Microsoft.ML.TensorFlow/TensorflowCatalog.cs [hunk 1]
```diff
@@ -35,5 +35,31 @@ public static class TensorflowCatalog
...
+        /// <param name="treatOutputAsBatched">If the first dimension of the output is unknown, should it be treated as batched or not.</param>
+        /// <example>
+        /// <format type="text/markdown">
+        /// <![CDATA[
+        /// [!code-csharp[LoadTensorFlowModel](~/../docs/samples/docs/samples/Microsoft.ML.Samples/Dynamic/TensorFlow/TextClassification.cs)]
+        /// ]]>
+        /// </format>
+        /// </example>
+        public static TensorFlowModel LoadTensorFlowModel(this ModelOperationsCatalog catalog, string modelLocation, bool treatOutputAsBatched)
+            => TensorFlowUtils.LoadTensorFlowModel(CatalogUtils.GetEnvironment(catalog), modelLocation, treatOutputAsBatched);
```
Comment (tarekgh) [inline src/Microsoft.ML.TensorFlow/TensorflowCatalog.cs:63]: Do we usually review the new APIs in the design review meeting? I am just wondering to understand the process of adding a new API if it is any different than anu other BCL library.

File: src/Microsoft.ML.TensorFlow/TensorflowCatalog.cs [hunk 2]
```diff
@@ -35,5 +35,31 @@ public static class TensorflowCatalog
...
+        /// <param name="modelLocation">Location of the TensorFlow model.</param>
+        /// <param name="treatOutputAsBatched">If the first dimension of the output is unknown, should it be treated as batched or not.</param>
+        /// <example>
+        /// <format type="text/markdown">
+        /// <![CDATA[
+        /// [!code-csharp[LoadTensorFlowModel](~/../docs/samples/docs/samples/Microsoft.ML.Samples/Dynamic/TensorFlow/TextClassification.cs)]
+        /// ]]>
+        /// </format>
+        /// </example>
+        public static TensorFlowModel LoadTensorFlowModel(this ModelOperationsCatalog catalog, string modelLocation, bool treatOutputAsBatched)
```
Comment (tarekgh) [inline src/Microsoft.ML.TensorFlow/TensorflowCatalog.cs:62]: Do we think in the future we may get more options than `treatOutputAsBatched`? I am asking to know if it will be better to have this as enum value instead of boolean?

File: src/Microsoft.ML.TensorFlow/TensorflowCatalog.cs (hunk 1 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.TensorFlow/TensorflowCatalog.cs:63]: I think that would be a good thing to start doing at some point, but we haven't had an official process for it before in ML.NET. At least all the new API's I have been involved in have just gone through a normal PR review like this, and sometimes a teams call if people have questions/concerns about it.

File: src/Microsoft.ML.TensorFlow/TensorflowCatalog.cs (hunk 2 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.TensorFlow/TensorflowCatalog.cs:62]: That is a good question that I don't know the answer too. From my understanding, this is the first change to these API's since there were first added, so my guess is that if things are added it wont be frequently, but I don't know if more will be added or not. 

I chose a bool this time because the other parameter dealing with batched input was a bool. I am happy to change it or discuss it more if you would like. I am not super familiar with the .NET guidelines, so it would be good for me to learn more.

File: src/Microsoft.ML.TensorFlow/TensorFlowModel.cs [hunk 3]
```diff
@@ -27,10 +28,12 @@ public sealed class TensorFlowModel : IDisposable
         /// <param name="env">An <see cref="IHostEnvironment"/> object.</param>
         /// <param name="session">TensorFlow session object.</param>
         /// <param name="modelLocation">Location of the model from where <paramref name="session"/> was loaded.</param>
-        internal TensorFlowModel(IHostEnvironment env, Session session, string modelLocation)
+        /// <param name="treatOutputAsBatched">If the first dimension of the output is unknown, should it be treated as batched or not.</param>
+        internal TensorFlowModel(IHostEnvironment env, Session session, string modelLocation, bool treatOutputAsBatched = true)
```
Comment (mstfbl) [inline src/Microsoft.ML.TensorFlow/TensorFlowModel.cs:32]: I'm wondering, what is you logic behind making `treatOutputAsBatched`'s default value to be `true`?

File: test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs [hunk 4]
```diff
@@ -1152,6 +1152,37 @@ public void TensorFlowGettingSchemaMultipleTimes()
...
+        [TensorFlowFact]
+        public void TreatOutputAsBatched()
+        {
+            MLContext mlContext = new MLContext();
+
+            var dataView = mlContext.Data.CreateTextLoader<TextInput>().Load(new MultiFileSource(null));
+
+            string modelLocation = @"model_string_test";
+
+            // When treatOutputAsBatched is defaultd to true, make sure the output is correct.
```
Comment (mstfbl) [inline test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs:1165]: Nit `defaulted`

File: src/Microsoft.ML.TensorFlow/TensorflowUtils.cs [hunk 5]
```diff
@@ -90,7 +90,16 @@ internal static DataViewSchema GetModelSchema(IExceptionContext ectx, Graph grap
                     DataViewType columnType = new VectorDataViewType(mlType);
                     if (!(Utils.Size(tensorShape) == 1 && tensorShape[0] <= 0) &&
                         (Utils.Size(tensorShape) > 0 && tensorShape.Skip(1).All(x => x > 0)))
-                        columnType = new VectorDataViewType(mlType, tensorShape[0] > 0 ? tensorShape : tensorShape.Skip(1).ToArray());
+                        if (treatOutputAsBatched)
```
Comment (mstfbl) [inline src/Microsoft.ML.TensorFlow/TensorflowUtils.cs:93]: It'd be good if you could add another more detailed explanation for this if/else here.

File: src/Microsoft.ML.TensorFlow/TensorflowCatalog.cs (hunk 2 above)
Comment (tarekgh) [reply src/Microsoft.ML.TensorFlow/TensorflowCatalog.cs:62]: Here is the guidlines for such case: https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/parameter-design#choosing-between-enum-and-boolean-parameters

I would prefer using enum here for future flexibility.

File: src/Microsoft.ML.TensorFlow/TensorFlowModel.cs (hunk 3 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.TensorFlow/TensorFlowModel.cs:32]: Most of the time we wanted the output to be batched and that is the current behavior that ML.NET has. So defaulting to true keeps the existing behavior but opens up the choice for the user to change it.

File: test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs (hunk 4 above)
Comment (michaelgsharp) [reply test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs:1165]: Good catch. Fixed.

File: src/Microsoft.ML.TensorFlow/TensorflowUtils.cs (hunk 5 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.TensorFlow/TensorflowUtils.cs:97]: Added more documentation to both the if and else.

File: src/Microsoft.ML.TensorFlow/TensorflowCatalog.cs (hunk 2 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.TensorFlow/TensorflowCatalog.cs:62]: So for the specific case of treating the output as batched or not, it will only ever be a yes or no answer. We can't do it half and half. Are the guidelines meaning that we want to be able to combine flags?

Comment (tarekgh) [reply src/Microsoft.ML.TensorFlow/TensorflowCatalog.cs:62]: It is not about combining flags. let's say you'll define the enum as 

```C#
public enum TensorflowCatalogOptions // Of course, you can choose better name
{
    None, 
    Batched
}
```

the API will be 

```C#
public static TensorFlowModel LoadTensorFlowModel(this ModelOperationsCatalog catalog, string modelLocation, TensorflowCatalogOptions options)
```

for any other method take this as defaulted parameter will do 


```C#
internal TensorFlowModel(IHostEnvironment env, Session session, string modelLocation, TensorflowCatalogOptions options = TensorflowCatalogOptions.Batched)
```

and in the implementation, you can test against TensorflowCatalogOptions.Batched value. 


Imagin in the future you will get a new case you want to support, say Filtered. which you can just add to the enum and handle it in the implmentation.

```C#
public enum TensorflowCatalogOptions // Of course, you can choose better name
{
    None, 
    Batched, 
    Filtered
}
```

It is up to you too if you want to make this enum as Flags enum which allow passing multiple values to this enum.

```C#
[Flags]
public enum TensorflowCatalogOptions // Of course, you can choose better name
{
    None = 0, 
    Batched = 1, 
    Filtered = 2
}
```

And that can allow passing multiple values if needed, like `TensorflowCatalogOptions.Batched | TensorflowCatalogOptions.Filtered`. And in the implmentation you can check specific case like `if (options & TensorflowCatalogOptions.Batched) != 0`.

In short, you need to think if this is possible in the future you can support more than options? and if it is possible multiple options can be supported together.

Comment (michaelgsharp) [reply src/Microsoft.ML.TensorFlow/TensorflowCatalog.cs:62]: So thinking about it, unless TensorFlow changes how they deal with their dimensions there won't be anything else that this could be. They only use ints to represent the dimensions so they are limited in how they express them, and essentially its that negative numbers mean unknown size and positive numbers mean known size. They would have to completely change how they represented the dimensions for that to change, which would then require a complete rewrite on our side to handle it if they did that. So I think we are safe to assume it will only be 2 options here, so I think we are safe to keep it as a boolean.

Comment (tarekgh) [reply src/Microsoft.ML.TensorFlow/TensorflowCatalog.cs:62]: You know better than me here :-) so I'll let you decide about it.

**Comment by michaelgsharp** [comment]
@eerhardt there are no TODOs in our TensorFlow code base. I'm also not seeing anything that pops out at me either when looking through the code base. I'm hoping that this means we are safe to assume there will be no more changes in the future.

File: src/Microsoft.ML.TensorFlow/TensorFlowModel.cs [hunk 6]
```diff
@@ -40,7 +43,7 @@ internal TensorFlowModel(IHostEnvironment env, Session session, string modelLoca
         /// </summary>
         public DataViewSchema GetModelSchema()
         {
-            return TensorFlowUtils.GetModelSchema(_env, Session.graph);
+            return TensorFlowUtils.GetModelSchema(_env, Session.graph, treatOutputAsBatched: TreatOutputAsBatched);
```
Comment (eerhardt) [inline src/Microsoft.ML.TensorFlow/TensorFlowModel.cs:46]: Nit: no need to use named parameter here, since the names are the same.

File: src/Microsoft.ML.TensorFlow/TensorflowUtils.cs [hunk 7]
```diff
@@ -32,7 +32,7 @@ internal static class TensorFlowUtils
         /// </summary>
         internal const string TensorflowUpstreamOperatorsKind = "TensorflowUpstreamOperators";
 
-        internal static DataViewSchema GetModelSchema(IExceptionContext ectx, Graph graph, string opType = null)
+        internal static DataViewSchema GetModelSchema(IExceptionContext ectx, Graph graph, string opType = null, bool treatOutputAsBatched = true)
```
Comment (eerhardt) [inline src/Microsoft.ML.TensorFlow/TensorflowUtils.cs:35]: Does anyone call this with the default parameter? Or can we just make it a normal parameter if all callers pass it in explicitly?

File: src/Microsoft.ML.TensorFlow/TensorflowUtils.cs [hunk 8]
```diff
@@ -90,7 +90,24 @@ internal static DataViewSchema GetModelSchema(IExceptionContext ectx, Graph grap
                     DataViewType columnType = new VectorDataViewType(mlType);
                     if (!(Utils.Size(tensorShape) == 1 && tensorShape[0] <= 0) &&
                         (Utils.Size(tensorShape) > 0 && tensorShape.Skip(1).All(x => x > 0)))
-                        columnType = new VectorDataViewType(mlType, tensorShape[0] > 0 ? tensorShape : tensorShape.Skip(1).ToArray());
+                        // When treatOutputAsBatched is true we keep the existing behaviour. This means that if the first dimension is greater
```
Comment (eerhardt) [inline src/Microsoft.ML.TensorFlow/TensorflowUtils.cs:93]: What does “keep the existing behavior” mean in this comment? When someone reads this in a year, what does “existing behavior” mean?

File: src/Microsoft.ML.TensorFlow/TensorflowUtils.cs [hunk 9]
```diff
@@ -90,7 +90,24 @@ internal static DataViewSchema GetModelSchema(IExceptionContext ectx, Graph grap
...
+                        // than 0 we take the tensor shape as is. If the first value is less then 0, we treat it as the batch input so we can
+                        // ignore it for the shape of the ML.NET vector. I.E. if the input dimensions are [-1, 5], ML.NET will read the -1 as
+                        // batch input, and so the ML.NET data type will be a vector of length 5.
+                        if (treatOutputAsBatched)
+                        {
+                            columnType = new VectorDataViewType(mlType, tensorShape[0] > 0 ? tensorShape : tensorShape.Skip(1).ToArray());
+                        }
+                        // When treatOutputAsBatched is false, if the first value is less than 0 we want to set it to 0. TensorFlow
+                        // represents and unkown size as -1, but ML.NET represents it as 0 so we need to convert it.
```
Comment (eerhardt) [inline src/Microsoft.ML.TensorFlow/TensorflowUtils.cs:102]: ‘and’ => ‘an’

File: test/Microsoft.ML.Tests/TensorFlowEstimatorTests.cs [hunk 10]
```diff
@@ -183,6 +183,52 @@ public void TestTensorFlow()
                 Assert.Equal(4, numRows);
             }
         }
+        
+        [TensorFlowFact]
+        public void TreatOutputAsBatched()
```
Comment (eerhardt) [inline test/Microsoft.ML.Tests/TensorFlowEstimatorTests.cs:188]: Do we have a test for saving and loading the model?

File: src/Microsoft.ML.TensorFlow/TensorflowUtils.cs (hunk 8 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.TensorFlow/TensorflowUtils.cs:93]: I was trying to explain what the existing behavior is with the remainder of the comment. Do you think that I should just remove that line?

File: src/Microsoft.ML.TensorFlow/TensorflowUtils.cs (hunk 9 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.TensorFlow/TensorflowUtils.cs:102]: Fixed.

File: test/Microsoft.ML.Tests/TensorFlowEstimatorTests.cs (hunk 10 above)
Comment (michaelgsharp) [reply test/Microsoft.ML.Tests/TensorFlowEstimatorTests.cs:188]: We do now. The `TestEstimatorCore()` method does a ton of tests, including saving and loading and making sure the schema/data are the same before/after the saving and loading.

File: src/Microsoft.ML.TensorFlow/TensorflowUtils.cs (hunk 7 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.TensorFlow/TensorflowUtils.cs:35]: I have refactored it so its now a normal parameter.

File: src/Microsoft.ML.TensorFlow/TensorFlowModel.cs (hunk 6 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.TensorFlow/TensorFlowModel.cs]: Fixed by making that parameter be a normal parameter, so the name is no longer needed.

File: src/Microsoft.ML.TensorFlow/TensorflowUtils.cs (hunk 8 above)
Comment (eerhardt) [reply src/Microsoft.ML.TensorFlow/TensorflowUtils.cs:93]: I think my issue is with "the existing behavior" clause. Once you check in this code, both cases are now "existing behavior".

So maybe just re-word it to
```suggestion
                        // treatOutputAsBatched == true means that if the first dimension is greater
```

Comment (michaelgsharp) [reply src/Microsoft.ML.TensorFlow/TensorflowUtils.cs:93]: That makes sense. Actually, I think you have made a similar comment on a prior PR that I did that "a year from now" no one will know what the "existing behavior" is. I'll try and remember that for the future as well.

File: test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs [hunk 11]
```diff
@@ -1227,11 +1226,13 @@ public void TensorFlowSentimentClassificationTest()
             // For explanation on how was the `sentiment_model` created 
             // c.f. https://github.com/dotnet/machinelearning-testdata/blob/master/Microsoft.ML.TensorFlow.TestModels/sentiment_model/README.md
             string modelLocation = @"sentiment_model";
-            using var pipelineModel = _mlContext.Model.LoadTensorFlowModel(modelLocation).ScoreTensorFlowModel(new[] { "Prediction/Softmax" }, new[] { "Features" })
+            using var pipelineModel = _mlContext.Model.LoadTensorFlowModel(modelLocation, false).ScoreTensorFlowModel(new[] { "Prediction/Softmax" }, new[] { "Features" })
                 .Append(_mlContext.Transforms.CopyColumns("Prediction", "Prediction/Softmax"))
                 .Fit(dataView);
             using var tfEnginePipe = _mlContext.Model.CreatePredictionEngine<TensorFlowSentiment, TensorFlowSentiment>(pipelineModel);
 
+            var schema = pipelineModel.GetOutputSchema(dataView.Schema);
```
Comment (eerhardt) [inline test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs:1234]: Should we do something with this `schema`? Can we assert something about it?

**Comment by eerhardt** [comment]
>         internal TensorFlowTransformer(IHostEnvironment env, TensorFlowModel tfModelInfo, string[] outputColumnNames, string[] inputColumnNames, bool addBatchDimensionInput = false)

Why doesn't this constructor get the new boolean parameter?

---
Refers to: src/Microsoft.ML.TensorFlow/TensorflowTransform.cs:113 in 63dabbc. [](commit_id = 63dabbcbcd31c6a40bf2cb89bae36d40c4060384, deletion_comment = False)

File: test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs (hunk 11 above)
Comment (michaelgsharp) [reply test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs:1234]: Actually, editing that test was an error on my part. Let me revert the changes I made to that test method. The only test that should be changed is `TreatOutputAsBatched`, and that one does have asserts on the schema checking the changes I made.

File: src/Microsoft.ML.TensorFlow/TensorflowTransform.cs [hunk 12]
```diff
@@ -876,6 +893,15 @@ internal sealed class Options : TransformInputBase
...
             [Argument(ArgumentType.AtMostOnce, HelpText = "Add a batch dimension to the input e.g. input = [224, 224, 3] => [-1, 224, 224, 3].", SortOrder = 16)]
             public bool AddBatchDimensionInputs = false;
+
+            /// <summary>
+            /// If the first dimension of the output is unknown, should it be treated as batched or not. e.g. output = [-1] will be read as a vector of unknown length when this is false.
+            /// </summary>
+            /// <remarks>
+            /// This parameter is used to deal with models that have unknown output shape and it needs to be interpreted in ML.NET as a vector of unkown length and not as a batch dimension.
+            /// </remarks>
+            [Argument(ArgumentType.AtMostOnce, HelpText = "If the first dimension of the output is unknown, should it be treated as batched or not. e.g. output = [-1] will be read as a vector of unkown length when this is false.", SortOrder = 17)]
```
Comment (eerhardt) [inline src/Microsoft.ML.TensorFlow/TensorflowTransform.cs:903]: `unkown` => `unknown`

repeated elsewhere

File: test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs (hunk 11 above)
Comment (michaelgsharp) [reply test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs]: Changes have been reverted, its now back to how it originally was.

File: src/Microsoft.ML.TensorFlow/TensorflowTransform.cs (hunk 12 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.TensorFlow/TensorflowTransform.cs:903]: Fixed.

**Comment by michaelgsharp** [comment]
@eerhardt it won't let me respond directly to your question about the constructor. It should be there. Nothing references that constructor so I must have missed it somehow. Its now fixed.

**Review by eerhardt — APPROVED** [review]
LGTM
